### PR TITLE
Proportional TSDB index stats for chunks partially fitting in queried time range.

### PR DESCRIPTION
**Pull request by salvacorts**
**What this PR does / why we need it**:

When computing index stats for a query, we sum up the stats from all the chunks the query would read, even if some chunks will only be read partially.

Moreover, if the query spans across two different TSDB tables, and there are chunks duplicated on both tables (at the end of the first one and at the beginning of the next one), the chunks' stats will be counted twice.

This PR improves how we compute stats by taking care of partially matching chunks. Assuming entries and bytes are evenly distributed in the chunk, we will take the proportional number of entries and the number of bytes that fit on the span.

For example, a chunk spans from t1 to t10 and has 10 entries and 10 KBs. A query from t3 through t7 would hit 50% of the chunk time span. Therefore, we estimate that the query would load 5 KBs and 5 entries from the chunk.

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/storage/stores/tsdb/index_client.go [hunk 1]
```diff
@@ -213,6 +213,7 @@ func (c *IndexClient) Stats(ctx context.Context, userID string, from, through mo
 			// start time earlier than start time of the table interval we are querying.
 			// for all other splits, consider only chunks that have from >= interval.Start
 			// so that we start after the start time of the index table we are querying.
+			// TODO: Maybe we don't need it anymore
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/storage/stores/tsdb/index_client.go]: Done

**Review by owen-d — APPROVED** [review]

**Review by cstyan — COMMENTED** [review]
> Moreover, if the query spans across two different TSDB tables, and there are chunks duplicated on both tables (at the end of the first one and at the beginning of the next one), the chunks' stats will be counted twice.

Not to derail things, but I'd like to understand how this scenario could happen. How could we end up with chunks that are present for two different tables (which I assume are actually different indicies? IE for different time ranges?)

**Comment by salvacorts** [comment]
> Not to derail things, but I'd like to understand how this scenario could happen. How could we end up with chunks that are present for two different tables (which I assume are actually different indicies? IE for different time ranges?)

I may have got it wrong. My understanding was that every 24h we would be creating a new index file. If a chunk was not full at the moment or cutting the new index, the chunk would be put on the index as it is, but the next index would also contain the same chunk.

If my understanding is not correct, do we have any docs explaining how it works in such case?

File: pkg/storage/stores/tsdb/compactor_test.go [hunk 2]
```diff
@@ -194,15 +194,19 @@ func buildStream(lbls labels.Labels, chunks index.ChunkMetas, userLabel string)
 }
 
 // buildChunkMetas builds 1ms wide chunk metas from -> to.
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/storage/stores/tsdb/compactor_test.go:196]: Good catch. I opened https://github.com/grafana/loki/pull/9124 fixing this.

**Pull request merged**